src/Graphing/Enrich.hs

```diff
@@ -0,0 +1,230 @@
+module Graphing.Enrich (
+  conanToArchives,
+) where
+
+import App.Fossa.LicenseScanner (licenseScanSourceUnit)
+import App.Fossa.VendoredDependency (VendoredDependency (..), VendoredDependencyScanMode (SkippingNotSupported))
+import App.Types (FullFileUploads (..))
+import Control.Algebra (Has)
+import Control.Carrier.Lift (Lift)
+import Control.Effect.Diagnostics (Diagnostics, ToDiagnostic, fatal)
+import Control.Effect.FossaApiClient (FossaApiClient)
+import Control.Effect.StickyLogger (StickyLogger)
+import Control.Monad (unless)
+import Data.Either (partitionEithers)
+import Data.List (find)
+import Data.List.NonEmpty (NonEmpty, nonEmpty, toList)
+import Data.List.NonEmpty qualified as NE
+import Data.Map (fromList)
+import Data.Map qualified as Map
+import Data.Maybe (listToMaybe)
+import Data.String.Conversion (toText)
+import Data.Text (Text, intercalate)
+import Data.Text.Extra (splitOnceOn)
+import DepTypes (DepType (ArchiveType, ConanType), Dependency (..), VerConstraint (CEq))
+import Diag.Diagnostic (ToDiagnostic (renderDiagnostic))
+import Effect.Exec (Exec)
+import Effect.Logger (Logger, indent, pretty, vsep)
+import Effect.ReadFS (ReadFS)
+import Graphing (Graphing, gmap, vertexList)
+import Path (Abs, Dir, Path)
+import Srclib.Converter (fetcherToDepType, toLocator, verConstraintToRevision)
+import Srclib.Types (Locator (locatorFetcher, locatorProject, locatorRevision))
+
+-- | Transform all 'Conan' dependency to 'Archive' dependency.
+-- It performs cli-side license scanning and upload of results. In this,
+-- transformation,
+-- -
+-- * It preserves edges of conan dependencies in the transformation.
+-- * TODO: Ignore non-production dependency in transformation
+conanToArchives ::
+  ( Has Diagnostics sig m
+  , Has (Lift IO) sig m
+  , Has StickyLogger sig m
+  , Has Logger sig m
+  , Has Exec sig m
+  , Has ReadFS sig m
+  , Has FossaApiClient sig m
+  ) =>
+  Path Abs Dir -> -- directory of the project's manifest
+  Graphing Dependency -> -- graph whose conan dependency to make archives
+  m (Graphing Dependency) -- graph with archive dependencies
+conanToArchives rootPath g =
+  -- If we do not have any conan dependencies, we do not
+  -- need to do any work!
+  case (null unableToTransformConanDep, transformedVendorDep) of
+    (True, Nothing) -> pure g
+    (False, _) -> fatal $ FailedToTransformConanDependency unableToTransformConanDep
+    (True, Just depsAndVendorDeps) -> do
+      let vendorDeps = NE.map snd depsAndVendorDeps
+
+      -- 1. We upload all vendor deps (transformed from conan deps),
+      -- since conan dependencies have to be scanned
+      -- before they can be referenced in graph as archive deps!
+      archiveLocators <-
+        licenseScanSourceUnit
+          -- always perform scan as vendor deps's
+          -- manifest are rarely updated with changes in practice
+          SkippingNotSupported
```

Comment:
I'm reading this as "in practice we often get code changes without a change in the version of the dependency"? If so, then yeah, this makes total sense.

Reply:
If, however, we're using the hash of the contents to get the version, then I think enabling skipping would be a nice performance boost for basically free